lib/bearerstrategy.js

```diff
@@ -118,121 +125,298 @@ const jws = require('jws');
+ * (1) for B2C, different policy may have different metadata, so the metadata should
+ * be loaded dynamically (and cached) in the authentication time based on the request
+ * (2) async issue, metadata loading is async, passport-http-bearer's `authenticate`
+ * function depends on the metadata, and it might be called before the metadata loading
+ * finishes
+ *
+ * Therefore, what we do here is let the metadata loading happen in `authenticate`
+ * function, and use waterfall to make sure the authentication code runs after the
+ * metadata loading is finished. To do this, we write our own `authenticate` function
+ * and no longer inherit passport-http-bearer strategy.
+ */
 function Strategy(options, verifyFn) {
-  const self = this;
-  const log = new Log('AzureAD: Bearer Strategy');
-  const verify = (typeof options === 'function') ? options : verifyFn;
-  const opts = (typeof options === 'function') ? {} : options;
+  passport.Strategy.call(this);
+  this.name = 'oauth-bearer'; // Me, a name I call myself.
+
+  this._verify = (typeof options === 'function') ? options : verifyFn;
+  this._options = (typeof options === 'function') ? {} : options;
+  this._passReqToCallback = this._options.passReqToCallback;
 
-  // Passport requires a verify function
-  if (typeof verify !== 'function') {
+  // (1) check the existence of the verify function
+  if (typeof this._verify !== 'function') {
     throw new TypeError('BearerStrategy requires a verify callback.');
   }
 
+  // (2) modify some fields of this._options
+  if (!this._options.realm)
+    this._options.realm = 'Users';
+  if (!Array.isArray(this._options.scope))
+    this._options.scope = [this._options.scope];
+
+  // (3) log some info of this._options
+
+  options = this._options;
+
   // if logging level specified, switch to it.
-  if (opts.loggingLevel) { log.levels('console', opts.loggingLevel); }
-
-  if (opts.policyName) {
-    log.info('B2C: We have been instructed that this is a B2C tenant. We will configure as required.');
-
-    if (!options.tenantName) {
-      throw new TypeError('BearerStrategy requires you pass the tenant name if using a B2C tenant.');
-    } else {
-      // We are replacing the common endpoint with the concrete metadata of a B2C tenant.
-      opts.identityMetadata = opts.identityMetadata
-        .replace('common', opts.tenantName)
-        .concat(`?p=${opts.policyName}`);
-    }
-  }
+  if (options.loggingLevel) { log.levels('console', options.loggingLevel); }
 
   // warn about validating the issuer
-  if (!opts.validateIssuer) {
+  if (!options.validateIssuer) {
     log.warn(`We are not validating the issuer.
       This is fine if you are expecting multiple organizations to connect to your app.
       Otherwise you should validate the issuer.`);
   }
 
   // if you want to check JWT audience (aud), provide a value here
-  if (opts.audience) {
-    log.info('Audience provided to Strategy was: ', opts.audience);
+  if (options.audience) {
```

Comment:
Checking audience is a mitigation against forwarding attacks. By default audience should always be validated.